File: TAC.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX 100

// Stack for operators
char stack[MAX];
int top = -1;

// Temp variable counter
int tempCount = 1;

// Stack functions
void push(char ch) {
    stack[++top] = ch;
}

char pop() {
    return stack[top--];
}

char peek() {
    return stack[top];
}

// Precedence
int precedence(char ch) {
    if (ch == '+' || ch == '-') return 1;
    if (ch == '*' || ch == '/') return 2;
    return 0;
}

// Generate temp variable
void newTemp(char *temp) {
    sprintf(temp, "t%d", tempCount++);
}

// Infix → Postfix
void infixToPostfix(char exp[], char postfix[]) {
    int k = 0;
    top = -1;

    for (int i = 0; i < strlen(exp); i++) {
        char ch = exp[i];

        if (isalnum(ch)) {
            postfix[k++] = ch;
        }
        else if (ch == '(') {
            push(ch);
        }
        else if (ch == ')') {
            while (top != -1 && peek() != '(') {
                postfix[k++] = pop();
            }
            pop();
        }
        else {
            while (top != -1 && precedence(peek()) >= precedence(ch)) {
                postfix[k++] = pop();
            }
            push(ch);
        }
    }

    while (top != -1) {
        postfix[k++] = pop();
    }

    postfix[k] = '\0';
}

// Generate TAC
void generateTAC(char postfix[], char lhs[]) {
    char st[MAX][MAX];
    int top2 = -1;

    for (int i = 0; i < strlen(postfix); i++) {
        char ch = postfix[i];

        if (isalnum(ch)) {
            char temp[2] = {ch, '\0'};
            strcpy(st[++top2], temp);
        }
        else {
            char op2[MAX], op1[MAX];
            strcpy(op2, st[top2--]);
            strcpy(op1, st[top2--]);

            char temp[MAX];
            newTemp(temp);

            printf("%s = %s %c %s\n", temp, op1, ch, op2);

            strcpy(st[++top2], temp);
        }
    }

    printf("%s = %s\n", lhs, st[top2]);
}

int main() {

    int n;
    char input[MAX], clean[MAX], lhs[MAX], rhs[MAX], postfix[MAX];

    printf("Enter number of expressions: ");
    scanf("%d", &n);
    getchar();

    for (int i = 0; i < n; i++) {

        printf("\nEnter expression (e.g., c=(a+b)*(c+d)):\n");
        fgets(input, sizeof(input), stdin);

        input[strcspn(input, "\n")] = '\0';

        // Remove spaces
        int k = 0;
        for (int j = 0; j < strlen(input); j++) {
            if (input[j] != ' ')
                clean[k++] = input[j];
        }
        clean[k] = '\0';

        // Split LHS and RHS
        char *token = strtok(clean, "=");
        strcpy(lhs, token);
        token = strtok(NULL, "=");
        strcpy(rhs, token);

        // Convert and generate TAC
        infixToPostfix(rhs, postfix);

        printf("Three Address Code:\n");
        generateTAC(postfix, lhs);
    }

    return 0;
}

File: TAC.java
import java.util.*;

class Main {

    static final int MAX = 100;

    // Stack for operators
    static Stack<Character> stack = new Stack<>();

    // Temp variable counter
    static int tempCount = 1;

    // Precedence
    static int precedence(char ch) {
        if (ch == '+' || ch == '-') return 1;
        if (ch == '*' || ch == '/') return 2;
        return 0;
    }

    // Generate temp variable
    static String newTemp() {
        return "t" + (tempCount++);
    }

    // Infix → Postfix
    static String infixToPostfix(String exp) {
        StringBuilder postfix = new StringBuilder();
        stack.clear();

        for (int i = 0; i < exp.length(); i++) {
            char ch = exp.charAt(i);

            if (Character.isLetterOrDigit(ch)) {
                postfix.append(ch);
            }
            else if (ch == '(') {
                stack.push(ch);
            }
            else if (ch == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    postfix.append(stack.pop());
                }
                if (!stack.isEmpty()) stack.pop();
            }
            else {
                while (!stack.isEmpty() && precedence(stack.peek()) >= precedence(ch)) {
                    postfix.append(stack.pop());
                }
                stack.push(ch);
            }
        }

        while (!stack.isEmpty()) {
            postfix.append(stack.pop());
        }

        return postfix.toString();
    }

    // Generate TAC
    static void generateTAC(String postfix, String lhs) {
        Stack<String> st = new Stack<>();

        for (int i = 0; i < postfix.length(); i++) {
            char ch = postfix.charAt(i);

            if (Character.isLetterOrDigit(ch)) {
                st.push(String.valueOf(ch));
            }
            else {
                String op2 = st.pop();
                String op1 = st.pop();

                String temp = newTemp();
                System.out.println(temp + " = " + op1 + " " + ch + " " + op2);

                st.push(temp);
            }
        }

        System.out.println(lhs + " = " + st.pop());
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of expressions: ");
        int n = sc.nextInt();
        sc.nextLine();

        for (int i = 0; i < n; i++) {

            System.out.println("\nEnter expression (e.g., c=(a+b)*(c+d)):");
            String input = sc.nextLine();

            // Remove spaces
            String clean = input.replaceAll(" ", "");

            // Split LHS and RHS
            String[] parts = clean.split("=");
            String lhs = parts[0];
            String rhs = parts[1];

            // Convert to postfix
            String postfix = infixToPostfix(rhs);

            // Generate TAC
            System.out.println("Three Address Code:");
            generateTAC(postfix, lhs);
        }

        sc.close();
    }
}